### ISSUE TITLE
MameDEV needs debugger enabled

### ISSUE THREAD

**Issue by jvook** [libretro/mame#30]
Mamedev would prefer that the debugger be accessible in Retroarch.  I believe this is for bug reporting purposes.  I've asked them to elaborate as necessary here.

**Comment by rb6502** [comment]
A few reasons:
1) It's used for advanced cheat finding
2) It's used by people who hack games (e.g. the gender-swap Donkey Kong that everyone loves)
3) It's used by home-brew developers for both the arcade platforms and the various computers and consoles (there's one guy that's developed several new VIC-20 games using MAME and the debugger to debug them)
4) It's used all the time by MAMEdev, so if you have any hopes of getting us to run RA or a libretro-compliant frontend, it needs to be included

And if you aren't running with it enabled there's no performance drop.

**Comment by r-type** [comment]
@jvook  @rb6502  yes it's a big lack in the current implementation , and we have to fill it.

for all you stated before and also because it's a upstream feature and this fork don't aim to be a castrated version (but a shallow fork with minimal changes to support the Libretro API backend , in my point of view ).
 
I'll look at try to find a way to deal with it.

**Comment by rb6502** [comment]
Thanks, I appreciate you guys looking into this.  I understand if it's not enabled for e.g. Android builds where it's not practical, but on the big 3 desktops it shouldn't be too hard.

**Comment by r-type** [comment]
I have done a little test to force build with  USE_QTDEBUG=1. (only tested in linux for now) 
it works the basis but I encounter 3 problems.

- first is trivial , have to change in retorach.cfg pause_nonactive = "false" , else retroarch will pause when it loose the focus , then  QT debug windows is not updated .

- second , more problematic , if a driver need to change FPS then libretro core raise an retro_get_system_av_info which re-init the internal libretro video driver (not the core one) and it failed to init . (just updating the geometry is fine ), I have to dig why !

- and the last but not the least , when exit from debuger or exit the core (choosing another game with mame osd works fine) , it crash at   debug_view_manager::free_view(debug_view&) ().
when close debugger , emu works but when I quit it segfault with  "QXcbEventReader" received signal SIGSEGV
I ve not dig too much here for now.

![mk1dbg](https://cloud.githubusercontent.com/assets/3614417/26467560/6daa2014-4193-11e7-8dfd-bedfc90126d5.png)
![x68kdbg](https://cloud.githubusercontent.com/assets/3614417/26468099/87ea3034-4195-11e7-839d-f8e4f0adb313.png)

**Comment by r-type** [comment]
Getting poc qt windows debugger working too , need to fix the exit debugger crash to go ahead. can take some time but I’m on...
I don’t have success with native windows debugger for now when activate it , i'm facing to 
"Value windows not supported for option debugger - falling back to auto"

on side note I get an retro ogl render module working but don't see any speed improvement over the soft one.

**Comment by r-type** [comment]
I'm slow but always on this issue ! 

I try to find an alternative to enable dbg on all platform without heavy dependency like Qt .
mamedev debugimgui seem to be a good compromise.

As bgfx is not supported (as it need lots of change in the upstream code to bake an libretro implementation ,and since there is little chance that this will happen) I made (quickly and badly) an  imgui software backend to permit debugimgui rendering.
it's not perfect but at least it show up some things. (bad rendering , bugs, some finishes...)

I block on a problem because initially wait_for_debugger waits for an action but the rendering does not yet appear (problem inherent to libretro implementation). Therefore I have no choice than force m_running to true. I hope find another way to deal this problem.

![dbgimgui_soft1](https://user-images.githubusercontent.com/3614417/27036988-e84f8ec0-4f86-11e7-962e-7c96631dc062.png)

![dbgimgui_soft0](https://user-images.githubusercontent.com/3614417/27037006-f0507274-4f86-11e7-979d-b005ce2f5b11.png)

**Comment by vadosnaprimer** [comment]
> I block on a problem because initially wait_for_debugger waits for an action but the rendering does not yet appear (problem inherent to libretro implementation).

You mean the part where `retro_load_game()` doesn't actually start up mame (that instead happens only in `retro_run()`)?
